### PR TITLE
Parse & Format

### PR DESCRIPTION
**Pull request by solomonhawk**
Addresses #13, #20, and #28

### CODE REVIEW

File: src/ca11y/index.js [hunk 1]
```diff
@@ -69,19 +146,47 @@ class Ca11y {
...
+  onInputChange(e) {
+    if (e.target.value) {
+      const result = this.props.parse(e.target.value)
+
+      if (result.valid) {
+        this.setDate(result.date)
+      } else {
+        // TODO: what does ca11y do when the user types garbage into the
+        // input. do we trigger an error handler? ignore the input and maintain
+        // previous state? reset the date entirely? reset it to today?
```
Comment (solomonhawk) [inline src/ca11y/index.js]: Interested in thoughts here.

File: src/ca11y/index.js [hunk 2]
```diff
@@ -69,19 +153,49 @@ class Ca11y {
...
+   **/
+  onInputChange(e) {
+    if (e.target.value) {
+      const result = this.props.parse(e.target.value)
+
+      if (result.valid) {
+        this.setDate(result.date)
+      } else {
+        this.setDate(today) // show current month
+        // set state to NOT SELECTED
```
Comment (solomonhawk) [inline src/ca11y/index.js]: Not quite done with this piece, but Dan and I decided how to handle the case where a user types invalid data into the input. We will be adding a UI state for "no date is currently selected".

**Comment by solomonhawk** [comment]
Hey jerk, you should update the README!

File: src/ca11y/index.js [hunk 3]
```diff
@@ -180,54 +317,43 @@ class Ca11y {
...
+    return template.monthHeader({ monthNameFull, monthName, fullYear })
   }
 
+  /**
+   * Renders the calendar. Updates the month header html content, the month caption
+   * html content, the calendar days html content, and the selected day DOM node reference.
+   * @return { Void }
+   **/
   render() {
+    // no super call because super `render` is a virtual method (not implemented in Base prototype)
```
Comment (solomonhawk) [inline src/ca11y/index.js]: :hocho:

File: src/ca11y/index.js [hunk 4]
```diff
@@ -235,6 +361,17 @@ class Ca11y {
   }
 }
 
+/**
+ * Ca11y.pickers is used to globally track instances of Ca11y within a give page.
```
Comment (solomonhawk) [inline src/ca11y/index.js:383]: :abc:

File: src/ca11y/index.js [hunk 5]
```diff
@@ -4,32 +4,93 @@ import zeroPad  from './lib/zeroPad'
 import keys     from './lib/keys'
```
Comment (solomonhawk) [inline src/ca11y/index.js:3]: :hocho: the `zeroPad` import, but leave it as a file available to commonjs consumers of this lib.

File: package.json [hunk 6]
```diff
@@ -53,6 +53,7 @@
     "webpack": "1.12.9"
   },
   "dependencies": {
-    "delegate": "git://github.com/vigetlabs/delegate#3.0.1"
+    "delegate": "git://github.com/vigetlabs/delegate#3.0.1",
```
Comment (nhunzaker) [inline package.json]: Looks like this was just merged upstream and published. Kudos to @greypants for PR'ing that:

https://github.com/zenorocha/delegate/commit/03474cc8f3b8650771f163a514bb6ee861080e19

Comment (nhunzaker) [inline package.json]: So possible to set this to "delegate: 3.0.1"?

File: src/ca11y/index.js (hunk 5 above)
Comment (nhunzaker) [inline src/ca11y/index.js:3]: What do you  mean?

Comment (nhunzaker) [inline src/ca11y/index.js:3]: Bizzare. Looks like some strange CSS jank:

screen shot 2016-01-13 at 12 47 36 pm

File: src/ca11y/index.js [hunk 7]
```diff
@@ -103,23 +229,47 @@ class Ca11y {
...
+   * @return { Void }
+   **/
+  selectDay(day, keepOpen=false, silent=false) {
+    const { month, fullYear } = this.state
+    const date = new Date(fullYear, month, day)
+
+    this.setDate(date, silent)
+
+    // update the input to reflect new state
+    if (!silent) this.ui.input.value = this.props.format(this.state.date)
```
Comment (nhunzaker) [inline src/ca11y/index.js]: Sort of an architectural question, but should this mutation go here or in a render/update method?

File: src/ca11y/index.js [hunk 8]
```diff
@@ -235,6 +361,17 @@ class Ca11y {
...
+/**
+ * Ca11y.pickers is used to globally track instances of Ca11y within a give page.
+ **/
 Ca11y.pickers = []
 
+/**
+ * Ca11y.init is used to instantiate multiple datepickers on a page simultaneously.
+ **/
+Ca11y.init = function(selector, options) {
+  const datePickers = Array.from(document.querySelectorAll(selector) || 0)
```
Comment (nhunzaker) [inline src/ca11y/index.js]: Is this polyfilled?

File: src/ca11y/index.js (hunk 5 above)
Comment (averyvery) [inline src/ca11y/index.js:3]: I think he's suggesting that, further up in the file, the zeroPad import should be removed (it's no longer used by the library, maybe?)

Comment (averyvery) [inline src/ca11y/index.js:3]: Yeah, looks like it's no longer used by anything. I think you should remove it, and the package...?

**Comment by averyvery** [comment]
:+1: really nice

File: src/ca11y/index.js (hunk 8 above)
Comment (averyvery) [inline src/ca11y/index.js]: Looks like no; you'll need this in your packagez/babel plugins: https://www.npmjs.com/package/babel-plugin-transform-array-from

Comment (nhunzaker) [inline src/ca11y/index.js]: Faaaaancyyy

File: package.json (hunk 6 above)
Comment (greypants) [inline package.json]: Yep! :tada: Make it so.

File: src/ca11y/index.js (hunk 8 above)
Comment (greypants) [inline src/ca11y/index.js]: Oh good catch... i thougth it was included in the es2015 preset, but looks like not.

Comment (nhunzaker) [inline src/ca11y/index.js]: Really rough... `Array.from` is supported in Firefox and Chrome, which are the only test targets we typically set up for Karma. We should explore setting up Browserstack to catch IE

File: src/ca11y/index.js [hunk 9]
```diff
@@ -4,32 +4,93 @@ import zeroPad  from './lib/zeroPad'
...
     this.props = Object.assign({}, defaults, options, { id: Ca11y.pickers.length })
     this.setInitialState()
     this.setUI(el)
-    this.selectDay(this.state.day, false, true)
     this.listen()
+    this.render()
+  }
+
+  /**
+   * An internal mechanism for updating state and re-rendering the UI.
```
Comment (solomonhawk) [inline src/ca11y/index.js:29]: Add `@param`s here.

File: src/ca11y/index.js [hunk 10]
```diff
@@ -43,22 +104,46 @@ class Ca11y {
...
       fullYear          : date.getFullYear(),
-      totalDays         : this.getDays(year, month),
-      firstWeekdayValue : this.getFirstDay(year, month)
+      totalDays         : util.getDays(year, month),
+      firstWeekdayValue : util.getFirstDay(year, month)
     }, silent)
   }
 
+  /**
+   * Initializes listeners. Initially set's Ca11y to have today's date selected.
```
Comment (solomonhawk) [inline src/ca11y/index.js:127]: This description is not accurate.

File: src/ca11y/index.js [hunk 11]
```diff
@@ -103,23 +229,47 @@ class Ca11y {
...
     this.selectDay(day)
   }
 
-  selectDay(day, keepOpen, silent) {
-    if (!silent) {
-      this.ui.input.value = `${zeroPad(this.state.monthDisplay)}/${zeroPad(day)}/${this.state.fullYear}`
-    }
-    this.setState({day})
+  /**
+   * Selects a day, updating internal state and, if `silent` is true, the input's
```
Comment (solomonhawk) [inline src/ca11y/index.js:257]: ... if silent is `false` ...

**Comment by tommymarshall** [comment]
:+1:

File: src/ca11y/index.js (hunk 4 above)
Comment (greypants) [inline src/ca11y/index.js:383]: abc wat?

File: src/ca11y/index.js (hunk 7 above)
Comment (greypants) [inline src/ca11y/index.js]: yes... moving it to render.

**Pull request merged**